state/charm.go

```diff
@@ -1115,7 +1115,15 @@ func (st *State) AddCharmMetadata(info CharmInfo) (*Charm, error) {
 		// Check if the charm doc already exists.
 		ch, err := st.findCharm(curl)
 		if errors.Is(err, errors.NotFound) {
-			return insertCharmOps(st, info)
+			ops, err := insertCharmOps(st, info)
+			// There is a race condition where the charm has been added between
+			// the call to findCharm and insertCharmOps. If the charm already
+			// exists, then continue.
+			if err != nil && !errors.Is(err, errors.AlreadyExists) {
```

Comment:
we can't continue, the charm was not loaded. We need to return jujutxn.ErrTransientFailure.

Reply:
Instead of AlreadyExists

Reply:
Ah I see the problem, though this change would not fix the race. 

If we get an already exists, we could do another `st.findCharm`. Would that be reasonable?

Reply:
Returning ErrTransientFailure causes the txn runner to retry the txn. Retrying the txn would mean the findCharm is called again. That should deal with the race.